### PR TITLE
Feature/fancy filter

### LINKED ISSUE #3: Store filter preferences of a user in locally accessible storage
```
Don't really think we need server side prefetching in the future though (Would like to store filter preferences in local storage for quicker results when revisiting without storing any user data in a db)
```

_Originally posted by @RopFoo in https://github.com/RopFoo/tonemato/pull/2#discussion_r1166057460_

### LINKED ISSUE #13: Create fancy filter - desktop
Create an air bnb like filter component to submit personalized requests to the teasers endpoint.

### LINKED ISSUE #34: Implement Skeleton Loader

### LINKED ISSUE #39: Fix asset locations (font and logo)
Vite is mad about some of the asset locations.

- fonts should be in /assets/fonts... instead of public/assets/fonts..
- logos (and other future images and svgs) should be imported from /src/... (hmr reasons)

### PR DESCRIPTION
**Pull request by ropfoo**
### This thing is ready to be reviewed ya

**🏁 Stuff on my todo list:**

- optimize styling (colors, hover states and stuff) -> wanna change that in a future issue
- [x] handle loading states -> accidentally also added skeleton loaders ☠️ (closes #34 )
- [x] add submit button (right now it submits on enter in last input field)
- [x] optimze some of the types
- [x] propper request parameter
- [x] fix weird default on empty input (all | bass | Ort)
- [x] local storage save  ->  closes #3 

closes #39

### CODE REVIEW

File: apps/smeargle/src/requests/fetch-teasers.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+const { VITE_RALTS_PORT, VITE_RALTS_DOMAIN } = import.meta.env;
+
+export async function getTeasers(params: TeaserRequestParams) {
+  const { instrument, category, zipCode } = params;
+  const query = `instrument=${instrument}&category=${category}&zipCode=${zipCode}`;
+
+  try {
+    const response = await fetch(
+      `${VITE_RALTS_DOMAIN}:${VITE_RALTS_PORT || 3005}/?${query}`
```
Comment (robinduerhager) [inline apps/smeargle/src/requests/fetch-teasers.ts]: Maybe we should create a utility function like `getBaseURL` which just concatenates the Domain with the Port.

**Comment by robinduerhager** [comment]
I think the `FilterSimple` should display the currently displayed filtered View, while `FilterDetailed` should incorporate already newly added, but maybe not already queried filter data. FilterSimple would then update, when the user clicks `search` so the view updates as well. I find it confusing if the FilterSimple changes without changing the filtered view because in the simple filter it might say `guitar` while `drums` results are presented.

filtersimple

**Comment by robinduerhager** [comment]
On initial load, clicking on `category` or `instrument` and then clicking on `location`, no cursor is displayed in the location field so it doesn't look active. Though, you can actually type in it. If you don't type something in it and just close the filter, the `Ort` prompt will disappear.

locationbug

**Comment by robinduerhager** [comment]
The `category` can only be activated by clicking on `Suche` but not by clicking on the category value like `Bands und Musiker`.

categorybug

**Comment by robinduerhager** [comment]
I think options of active input fields should be displayed even if the value has been written out so you can be sure that the input value is actually valid. Alternatively some other form of giving the user some validation that his input is a known category could be fine as well.

Also i think it would be nice if `Alle` would be more like a prompt if i didn't select it before. It would be nice if i could just start typing the name of the instrument that i'm playing instead of first deleting the word `Alle`.

instrumentbug

**Comment by robinduerhager** [comment]
All types look like real values though the styling is different since the `location` value is a prompt and the `instrument` value is a string. Though i think a consistent use of input types would be beneficial here, else this could be a bit confusing for the user.

filtersimple

**Comment by robinduerhager** [comment]
I don't think we have to fix everything in this PR. Maybe converting these into smaller issues would be good :D

**Comment by ropfoo** [comment]
Yep let's focus on actual bugs in this PR and put all the rest into issue :) 

Didn't test it in Firefox btw so I think some of the on click issue might come from there 

frontend ist mein Lieblingsfach

**Comment by ropfoo** [comment]
Nice qa btw :p

**Comment by ropfoo** [comment]
Also didn't worry about Accessibility tbh so this is also a future issue to keep in mind

File: apps/smeargle/src/components/Filter/FilterDetailed/useFilterStorage.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+  const storeFilterState = () => {
+    localStorage.setItem('filter', JSON.stringify(filterState.filter));
+  };
+
+  const getFilterStorage = () => {
+    const filterStorage = localStorage.getItem('filter');
+    return filterStorage
+      ? (JSON.parse(filterStorage) as FilterState['filter'])
+      : null;
```
Comment (robinduerhager) [inline apps/smeargle/src/components/Filter/FilterDetailed/useFilterStorage.ts]:
```suggestion
    return filterStorage && (JSON.parse(filterStorage) as FilterState['filter']);
```

**Comment by robinduerhager** [comment]
> Nice qa btw :p

nodgif

**Comment by robinduerhager** [comment]
> Didn't test it in Firefox btw so I think some of the on click issue might come from there

Oh yeah Firefox is sometimes... weird :D.

> Yep let's focus on actual bugs in this PR and put all the rest into issue :)

Yeah i just wanted to aggregate a lot of things which pointed out to me. Luckily GitHub has this handy feature to convert comments to issues :D.

> frontend ist mein Lieblingsfach

cssgif

File: apps/smeargle/src/components/Filter/FilterDetailed/useComboBoxInput.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    if (props.value) {
+      updateValue(props.name, props.value);
+    }
+  });
+
+  const isInputInOptions = () => {
+    if (props.isDropdown) return false;
+    if (!props.options) return false;
+    return (
+      Object.entries(props.options).filter(([, text]) => text === inputValue())
```
Comment (robinduerhager) [inline apps/smeargle/src/components/Filter/FilterDetailed/useComboBoxInput.ts:31]: Sometimes we have `[_, otherVar]` and sometimes `[, otherVar]`. Also in `useFilterDetailed.ts` but i can't add a suggestion in this, lel.
```suggestion
      Object.entries(props.options).filter(([_, text]) => text === inputValue())
```

File: apps/smeargle/src/components/Filter/FilterDetailed/index.tsx [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        'dark:bg-joplin border-1 dark:border-janis border-elvis dark:shadow-filter-dark shadow-filter-light bg-snow absolute flex h-20 w-fit translate-y-[55px] items-center rounded-full transition-all duration-300'
+      )}
+    >
+      <CategorySelector value={filterState.filter.category?.value ?? 'all'} />
+      <div class="mr-2" />
+      <InstrumentSelector
+        value={filterState.filter.instrument?.value ?? 'all'}
+      />
+      <div class="mr-2" />
+      <ZipCodeField value={filterState.filter.zipCode?.value ?? ''} />
```
Comment (robinduerhager) [inline apps/smeargle/src/components/Filter/FilterDetailed/index.tsx]: I think using `||` could be better here according to [this post](https://stackoverflow.com/questions/70334826/difference-between-and-in-javascript).
```suggestion
      <CategorySelector value={filterState.filter.category?.value || 'all'} />
      <div class="mr-2" />
      <InstrumentSelector
        value={filterState.filter.instrument?.value || 'all'}
      />
      <div class="mr-2" />
      <ZipCodeField value={filterState.filter.zipCode?.value || ''} />
```

File: apps/smeargle/src/components/Filter/FilterDetailed/FilterComboBox.tsx [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    inputValue,
+    handleInputKeyDown,
+    selectOption,
+    isInputInOptions,
+    handleBlur,
+  } = useComboBoxInput(props);
+
+  const [filterState, { setActive }] = useFilterContext();
+
+  const [, { isLoading }] = useTeasers();
```
Comment (robinduerhager) [inline apps/smeargle/src/components/Filter/FilterDetailed/FilterComboBox.tsx:26]: Sometimes we have [_, otherVar] and sometimes [, otherVar]. Also in useFilterDetailed.ts but i can't add a suggestion in this, lel.
```suggestion
  const [_, { isLoading }] = useTeasers();
```

**Review by robinduerhager — CHANGES_REQUESTED** [review]
reviewed

File: apps/smeargle/src/components/Filter/FilterDetailed/useComboBoxInput.ts (hunk 3 above)
Comment (ropfoo) [reply apps/smeargle/src/components/Filter/FilterDetailed/useComboBoxInput.ts:31]: jup thats a default in other languages but ts is mad if I reuse _ as a variable... just gonna remove them

File: apps/smeargle/src/components/Filter/FilterDetailed/index.tsx (hunk 4 above)
Comment (ropfoo) [reply apps/smeargle/src/components/Filter/FilterDetailed/index.tsx]: uhhhm this post is about the && and ?? operator... So I want the right side if the left side is null or undefined... || is for checking falsy values (including 0) but I dont really see the use case here. In fact I think we should allow empty strings as propper values as an alternative to "all". But maybe you need to elaborate more on this :D

**Comment by ropfoo** [comment]
Stuff I fixed:

- empty zip code default
- auto select next filter value
- focus and blur issues 
-> [this](https://github.com/ropfoo/tonemato/pull/43#issuecomment-1537167007)
-> [and this](https://github.com/ropfoo/tonemato/pull/43#issuecomment-1537167304)
- [keep suggestions open on matching input + little highlight](https://github.com/ropfoo/tonemato/pull/43#issuecomment-1537168146)

Stuff I did not fix

- [i dont know -maybe future issue?](https://github.com/ropfoo/tonemato/pull/43#issuecomment-1537168849)
- [did not really understand the issue - future issue? :D ](https://github.com/ropfoo/tonemato/pull/43#issuecomment-1537166393)

maybe some of the ux ui stuff we should rather discuss in person 

[englisch ist mein lieblingsfach](https://www.youtube.com/watch?v=39UDZMgPg5k&ab_channel=Levon)

File: apps/smeargle/src/components/Filter/FilterDetailed/index.tsx (hunk 4 above)
Comment (robinduerhager) [reply apps/smeargle/src/components/Filter/FilterDetailed/index.tsx]: Nah it also shows how `||` behaves (god, JS is such a mess :D).
With `||` you can set default values a little more expressively. E.g.: At the category selector when using `??`, `false` or `0` as boolean and number values respectively are possible inputs which could lead to more potential errors i think.

```md
                        |    ||   |    &&   |    ??   |
    --------------------+---------+---------+---------+
        0, 'default'    |'default'|    0    |    0    |
       '', 'default'    |'default'|   ''    |   ''    |
    false, 'default'    |'default'|  false  |  false  |
     null, 'default'    |'default'|   null  |'default'|
undefined, 'default'    |'default'|undefined|'default'|
     true, 'default'    |  true   |'default'|  true   |
       20, 'default'    |   20    |'default'|   20    |
```

**Comment by ropfoo** [comment]
🔮 Features for future issues:

- propper validation (zip code - maybe using [zod](https://github.com/colinhacks/zod))
- more "mature" ux ui
- mobile support
- better request handling (when backend is ready)
- improve accessability

**Pull request closed without merging**